components/brave_wallet/common/bitcoin_utils.cc

```diff
@@ -0,0 +1,101 @@
+
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/common/bitcoin_utils.h"
+
+#include <utility>
+
+#include "brave/components/brave_wallet/common/hash_utils.h"
+#include "brave/third_party/bitcoin-core/src/src/bech32.h"
+#include "brave/third_party/bitcoin-core/src/src/util/strencodings.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace {
+
+constexpr size_t kBech32WitnessMaxLength = 40;
+constexpr size_t kBech32WitnessMinLength = 2;
+constexpr size_t kP2WPKHLength = 20;
+constexpr size_t kP2WSHLength = 32;
+
+}  // namespace
+
+namespace brave_wallet {
+
+DecodedBitcoinAddress::DecodedBitcoinAddress() = default;
+DecodedBitcoinAddress::~DecodedBitcoinAddress() = default;
+DecodedBitcoinAddress::DecodedBitcoinAddress(
+    const DecodedBitcoinAddress& other) = default;
+DecodedBitcoinAddress& DecodedBitcoinAddress::operator=(
+    const DecodedBitcoinAddress& other) = default;
+DecodedBitcoinAddress::DecodedBitcoinAddress(DecodedBitcoinAddress&& other) =
+    default;
+DecodedBitcoinAddress& DecodedBitcoinAddress::operator=(
+    DecodedBitcoinAddress&& other) = default;
+
+// TODO(apaymyshev): should also return decoded address kind(pubkey hash, script
+// hash, etc).
+absl::optional<DecodedBitcoinAddress> DecodeBitcoinAddress(
+    const std::string& address,
+    bool testnet) {
+  // TODO(apaymyshev): support legacy addresses.
+  auto bech_result = bech32::Decode(address);
+
+  // https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#segwit-address-format
+  if (bech_result.second.empty() ||
+      bech_result.first != (testnet ? "tb" : "bc")) {
+    return absl::nullopt;
+  }
+
+  DecodedBitcoinAddress result;
+  if (bech_result.second[0] > 16) {
+    return absl::nullopt;
+  }
+
+  std::vector<uint8_t> data;
+  data.reserve(((bech_result.second.size() - 1) * 5) / 8);
+  if (!ConvertBits<5, 8, false>([&](unsigned char c) { data.push_back(c); },
+                                bech_result.second.begin() + 1,
+                                bech_result.second.end())) {
+    return absl::nullopt;
+  }
+
+  result.witness_version = bech_result.second[0];
+  result.pubkey_hash = std::move(data);
+
+  // https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#witness-program
+  if (result.witness_version == 0) {
+    if (result.pubkey_hash.size() == kP2WPKHLength) {
+      result.address_type = BitcoinAddressType::kWitnessV0PubkeyHash;
+    } else if (result.pubkey_hash.size() == kP2WSHLength) {
+      result.address_type = BitcoinAddressType::kWitnessV0ScriptHash;
+    } else {
+      return absl::nullopt;
+    }
+  }
```

Comment:
We didn't handle version 1-16 so the `address_type` will be undefined, shouldn't we assign it to `kWitnessUnknown` in `else` case or just make `kWitnessUnknown` the default value?

And we need a test case for it.

Reply:
Fixed https://github.com/brave/brave-core/pull/17978/commits/2f117c534c8236292c2bbc209de5195830ccb9c5